daml/ContingentClaims/Recursion.daml

```diff
@@ -0,0 +1,73 @@
+--
+-- Copyright (c) 2021, Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+--
+
+{-# OPTIONS -Wno-deprecations #-} -- To supress 'Monad' warnings
+
+module ContingentClaims.Recursion (
+    cataM
+  , paraM
+  , anaM
+  , apoM
+  , apoCataM
+  , funzip
+  , synthesize
+  , inherit
+  , subTreeSize'
+) where
+
+import Prelude hiding (sequence, mapA, sum)
+import DA.Traversable (Traversable, sequence, mapA)
+import Daml.Control.Recursion (Recursive(..), Corecursive(..), Cofree(..))
+import DA.Foldable (Foldable, sum)
+import Daml.Control.Arrow ((&&&), (|||), (>>>), (<<<), Kleisli(..))
+
+-- The morphisms ending in 'M' are monadic variants, allowing to interleave e.g. `Update` or `Script`.
+-- `cataM` After Tim Williams' talk, https://www.youtube.com/watch?v=Zw9KeP3OzpU.
+
+-- TODO: the two folds can probably be simplified with Traverse.mapA too.
+cataM : (Monad m, Traversable f, Recursive b f) => (f a -> m a) -> b -> m a
+cataM f b = (project >>> fmap (cataM f) >>> (>>= f) . sequence) b
+
+paraM : (Monad m, Traversable f, Recursive b f) => (f (b, a) -> m a) -> b -> m a
+paraM f b = (project >>> fmap (runKleisli $ Kleisli pure &&& Kleisli (paraM f)) >>> (>>= f) . sequence) b
+
+-- FIXME : I think the algebra should be a -> m (f a)
+anaM : (Monad m, Traversable f, Corecursive b f) => (a -> f (m a)) -> a -> m b
+anaM f a = (fmap embed <<< mapA (>>= anaM f) <<< f) a
+
+apoM : (Monad m, Traversable f, Corecursive b f) => (a -> m (f (Either b a))) -> a -> m b
+apoM f a = (fmap embed <<< (>>= (mapA (pure ||| apoM f))) <<< f) a
+
+-- | Specialised lazy re-fold, used by `lifecycle`.
+apoCataM : (Monad m, Traversable f, Corecursive b f) => (f b -> b) -> (a -> m (f (Either b a))) -> a -> m b
+apoCataM g f a = _apoCataM g f a 
+
+-- | HIDE
+_apoCataM g f a = (fmap g <<< (>>= (mapA (pure ||| apoCataM g f))) <<< f) a
```

Comment:
why is this helper function needed?

Reply:
It's not. I think it was left-over from something different I was trying to implement. :+1: